docs/design/auditing-containment.md

```diff
@@ -0,0 +1,205 @@
+The contained flag should only be relevant to the audit service.
+
+## Design
+
+### Identifying Nodes That Need to Be Contained
+In the audit verifier, we need a better system for handling the different cases in which a storage node may not respond to an audit request.
+Here are a few possibile cases:
+1. The node is busy fulfilling other requests and can't respond to the audit request within the audit timeout defined on the Satellite.
+2. The node does not have the audit piece anymore because the SNO deleted it.
+3. The node can't read the piece due to a file permission issue (a SNO config mistake).
+4. The node can't read the piece due to bad sectors on the HDD.
+
+For 1, the node doesn't necessarily need to be contained. For 2, the node does need to be contained.
+For 3 and 4, the node should probably also be contained.
+
+We need to find out what the error messages are for these different cases so that we don't lump the first case with the rest.
+
+### Handling Contained Nodes
+New functionality needs to be added to the audit verifier, where upon receiving a pointer, the verifier should first iterate through the nodes listed in the pointer and check if they have a `contained` field set to true on their NodeDossier or not.
+
+For nodes that are not contained, the audit will continue as normal.
+
+For nodes that are contained, the verifier will ask the ContainmentDB for the hash of the erasure share expected from the node when it was originally audited.
+
+The verifier will then attempt to download that erasure share from the contained node.
+From there, the following cases can occur:
+- Passes audit ->  `pendingAudit` is deleted from the ContainmentDB  -> stats updated in overlaycache/satelliteDB
+- Fails audit -> `pendingAudit` is deleted from the ContainmentDB -> stats updated in overlaycache/satelliteDB
+- Refuses to respond again -> the `pendingAudit`'s `reverifyCount` is incremented in the ContainmentDB ->
+    - -> If that count exceeds the reverify limit, then the node’s stats will be updated to reflect an audit failure and the node will be removed from the ContainmentDB.
+    - -> If the count does not exceed the reverify limit, the `pendingAudit` will remain in the ContainmentDB.
+
+```go
+type pendingAudit struct {
+    nodeID            storj.NodeID
+    pieceID           storj.PieceID
+    pieceNum          int
+    stripeIndex       int
+    shareSize         int64
+    expectedShareHash []byte
+    reverifyCount     int // number of times reverify has been attempted
+}
+```
+`pendingAudit` entries are created and added to the ContainmentDB when a node seems to be online and opens a connection with the satellite to be audited (this happens within the existing Verify function in the audit package), but then refuses to send the requested erasure share.
+
+Additionally, the satellite should make sure to empty the ContainmentDB of certain pending audits when segments are deleted.
+
+```go
+type ContainmentDB interface {
```

Comment:
Shouldn't there be a `Create` function as well?

Reply:
I think @wthorp suggested that `IncrementPending` (or `IncrementFail`) should be used for both creating and updating `pendingAudit` entries.